rpk: integrate profiles with `rpk container`

r-vasquez opened the pull request:
This PR introduces a new behaviour to `rpk container`:

Now, when the user creates a container using `rpk container start`, rpk will create and use a profile with the kafka broker and admin API and schema registry addresses so the user can start using rpk seamlessly without having to worry about the flags nor the environment variables

### Examples
```
$ rpk container start -n 3
Checking for a local image...
Creating network redpanda
Starting cluster...
Waiting for the cluster to be ready...

Cluster started!
NODE-ID  STATUS   KAFKA-ADDRESS    ADMIN-ADDRESS    PROXY-ADDRESS
0        running  127.0.0.1:36131  127.0.0.1:46387  127.0.0.1:44137
1        running  127.0.0.1:38361  127.0.0.1:44093  127.0.0.1:42481
2        running  127.0.0.1:40911  127.0.0.1:37757  127.0.0.1:36899

Created "rpk-container" profile.

You can use rpk to interact with this cluster. E.g:

    rpk cluster info
    rpk cluster health

$ rpk cluster info 
CLUSTER
=======
redpanda.feb945f2-8bb7-4212-b0a6-80aee86b7d35

BROKERS
=======
ID    HOST       PORT
0*    127.0.0.1  36131
1     127.0.0.1  38361
2     127.0.0.1  40911
```

We also delete the profile when the user deletes the cluster:

```
$ rpk container purge
Stopping node 0
Stopping node 2
Stopping node 1
Removed container rp-node-2 (node 2)
Removed container rp-node-1 (node 1)
Removed container rp-node-0 (node 0)
Deleted cluster data.
Deleted "rpk-container" profile.
This was the selected profile; rpk will use defaults until a new profile is selected or a new container is created.
```

When querying the cluster status, we chose to display the flags based on the current profile: if the profile is not the rpk-container profile, then we display the flags:

```
$ rpk profile current
default
$rpk container status
NODE-ID  STATUS   KAFKA-ADDRESS    ADMIN-ADDRESS    PROXY-ADDRESS
0        running  127.0.0.1:45969  127.0.0.1:46365  127.0.0.1:37121

You can use rpk to interact with this cluster. E.g:

    rpk cluster info -X brokers=127.0.0.1:45969
    rpk cluster health -X admin.hosts=127.0.0.1:46365

You may also set an environment variable with the comma-separated list of
broker and admin API addresses:

    export RPK_BROKERS="127.0.0.1:45969"
    export RPK_ADMIN_HOSTS="127.0.0.1:46365"
    rpk cluster info
    rpk cluster health

$ rpk profile use rpk-container 
Set current profile to "rpk-container".

$ rpk container status         
NODE-ID  STATUS   KAFKA-ADDRESS    ADMIN-ADDRESS    PROXY-ADDRESS
0        running  127.0.0.1:45969  127.0.0.1:46365  127.0.0.1:37121

You can use rpk to interact with this cluster. E.g:

    rpk cluster info
    rpk cluster health
``` 

Lastly, this PR also includes a way to create a profile from a running cluster created using rpk container:

```
$ rpk profile create --from-rpk-container 
Created and switched to new profile "rpk-container".

$ rpk profile print
name: rpk-container
description: Automatically generated profile from running 'rpk container start'
kafka_api:
    brokers:
        - 127.0.0.1:45969
admin_api:
    addresses:
        - 127.0.0.1:46365
schema_registry:
    addresses:
        - 127.0.0.1:45479
```
## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

### Improvements

* rpk now will create a profile when you run `rpk container create`.

twmb commented on src/go/rpk/pkg/cli/container/common/common.go:
```diff
@@ -459,3 +469,44 @@ This can happen for a couple of reasons:
+
+	profile := config.RpkProfile{
+		Name:        ContainerProfileName,
+		Description: "Automatically generated profile from running 'rpk container start'",
```
s/from running/from

twmb commented on src/go/rpk/pkg/cli/container/common/common.go:
```diff
@@ -459,3 +469,44 @@ This can happen for a couple of reasons:
+	return nil
+}
+
+const ContainerProfileName = "rpk-container"
+
+var ErrContainerProfileExists = fmt.Errorf("%q profile already exists", ContainerProfileName)
```
Maybe doc? Although this package isn't exemplary w.r.t. doc comments at the moment.

twmb commented on src/go/rpk/pkg/cli/container/purge.go:
```diff
@@ -12,43 +12,69 @@ package container
+				fmt.Fprintf(os.Stderr, "unable to delete %q profile: %v; you may delete the profile manually running 'rpk profile delete %v'", common.ContainerProfileName, err, common.ContainerProfileName)
+				return
+			}
+			fmt.Printf("Deleted %q profile.\n", common.ContainerProfileName)
```
move `profile` in front of `%q` here -- this will also match the `rpk profile delete` command output.

twmb commented on src/go/rpk/pkg/cli/profile/create.go:
```diff
@@ -131,6 +141,7 @@ func createProfile(
 	fromRedpanda string,
```
fwiw I wrote this function terribly to begin with but, it does the job. Originally I had more unit testing ambitions.

twmb commented on src/go/rpk/pkg/cli/container/start.go:
```diff
@@ -66,60 +68,51 @@ func newStartCommand() *cobra.Command {
+				renderClusterInteract(dockerNodes, true)
+				return
+			}
+			if errors.Is(err, common.ErrContainerProfileExists) {
+				fmt.Printf("Unable to create a profile for the rpk container: %v; you may delete it and create a new one running 'rpk profile create --from-rpk-container'\n", err)
```
We need to say that the name is here since we don't output it anywhere really. What do you think about,

```
Unable to create a profile for the rpk container: %v.

You can retry profile creation by running:
    rpk profile delete %s; rpk profile create --from-rpk-container
```
and then using err, common.ContainerProfileName as args
(it's a bit long but at least it's a bit less wide?)

twmb commented on src/go/rpk/pkg/cli/container/start.go:
```diff
@@ -66,60 +68,51 @@ func newStartCommand() *cobra.Command {
+	command.Flags().UintVar(&retries, "retries", 10, "The amount of times to check for the cluster before considering it unstable and exiting")
+	command.Flags().StringVar(&image, "image", common.DefaultImage(), "An arbitrary container image to use")
+	command.Flags().BoolVar(&pull, "pull", false, "Force pull the container image used")
+	command.Flags().BoolVar(&noProfile, "no-profile", false, "If true, rpk will not create an rpk profile after creating a cluster")
```
Do we want to use no-profile as a global flag to disable using profiles? Or should we use something else?

r-vasquez replied:
:thinking: Maybe use here `--no-profile-creation`.

A little bit longer but I don't expect this flag to have a lot of usage tbh, and this way we leave the room open for a global `--no-profile`

wdyt?

r-vasquez commented:
[Last force push](https://github.com/redpanda-data/redpanda/compare/b93aba488ce459fbd9a233a1d93d35a277e651d7..209ed9bec205f2e7df571513d153e841f721462f):

- Addressing PR comments.
